### PR TITLE
Automatic detection of CPU capabilities

### PR DESCRIPTION
**Pull request by tevador**
Can autodetect AES, SSSE3 and AVX2 and sets the JIT flag if supported.

Usage:
```c
randomx_flags flags = randomx_get_flags();
randomx_cache *myCache = randomx_alloc_cache(flags);
randomx_init_cache(myCache, &myKey, sizeof myKey);
randomx_vm *myMachine = randomx_create_vm(flags, myCache, NULL);
``` 

AES detection on ARM64 needs testing.

### CODE REVIEW

**Comment by tevador** [comment]
BTW also includes `--auto` flag in the benchmark, so:
```
./randomx-benchmark --mine --jit --largePages --init 8 --avx2
```
can be simplified to
```
./randomx-benchmark --mine --auto --largePages
```

**Comment by tevador** [comment]
@SChernykh please review

File: src/randomx.cpp [hunk 1]
```diff
@@ -33,18 +33,38 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
...
+			flags |= RANDOMX_FLAG_ARGON2_SSSE3;
+		}
+		return flags;
+	}
+
 	randomx_cache *randomx_alloc_cache(randomx_flags flags) {
 		randomx_cache *cache = nullptr;
+		auto impl = randomx::selectArgonImpl(flags);
+		if (impl == nullptr) {
+			return cache;
```
Comment (SChernykh) [inline src/randomx.cpp]: It's probably a bad idea to return uninitialized cache here even if impl can't be nullptr with current code. It should output some error and abort, or throw an exception.

File: src/randomx.cpp [hunk 2]
```diff
@@ -33,18 +33,38 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
...
+#include "cpu.hpp"
 #include <cassert>
 #include <limits>
 
 extern "C" {
 
+	randomx_flags randomx_get_flags() {
+		randomx_flags flags = RANDOMX_HAVE_COMPILER ? RANDOMX_FLAG_JIT : RANDOMX_FLAG_DEFAULT;
+		randomx::Cpu cpu;
+		if (HAVE_AES && cpu.hasAes()) {
```
Comment (SChernykh) [inline src/randomx.cpp]: It's better to wrap it into `#if HAVE_AES == 1 ... #endif`, no need to make it compile-time when preprocessor can handle it.

File: src/randomx.cpp (hunk 1 above)
Comment (tevador) [reply src/randomx.cpp]: 1. `impl` can be null if the caller selects an unsupported implementation (e.g. AVX2 on ARM).
1. It's not returning an uninitialized cache but simply a NULL pointer. It's up to the client to check for NULLs.
1. This is a C-language API, so we can't throw exceptions unless we handle them internally.

Comment (SChernykh) [reply src/randomx.cpp]: Then it's ok, I somehow thought cache was already allocated at this point.

**Pull request merged**